xds/internal/resolver/watch_service_test.go

```diff
@@ -16,375 +16,375 @@
+	r, err := builder.Build(target, tcc, resolver.BuildOptions{})
+	if err != nil {
+		t.Fatalf("Failed to build xDS resolver for target %q: %v", target, err)
 	}
+	t.Cleanup(r.Close)
+	return tcc, r
 }
 
-// TestServiceWatchLDSUpdate covers the case that after first LDS and first RDS
-// response, the second LDS response trigger an new RDS watch, and an update of
-// the old RDS watch doesn't trigger update to service callback.
-func (s) TestServiceWatchLDSUpdate(t *testing.T) {
-	serviceUpdateCh := testutils.NewChannel()
-	xdsC := fakeclient.NewClient()
-	cancelWatch := watchService(xdsC, targetStr, func(update serviceUpdate, err error) {
-		serviceUpdateCh.Send(serviceUpdateErr{u: update, err: err})
-	}, nil)
-	defer cancelWatch()
-
-	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
-	defer cancel()
-	waitForWatchListener(ctx, t, xdsC, targetStr)
-	xdsC.InvokeWatchListenerCallback(xdsresource.ListenerUpdate{RouteConfigName: routeStr}, nil)
-	waitForWatchRouteConfig(ctx, t, xdsC, routeStr)
-
-	wantUpdate := serviceUpdate{virtualHost: &xdsresource.VirtualHost{Domains: []string{"target"}, Routes: []*xdsresource.Route{{Prefix: newStringP(""), WeightedClusters: map[string]xdsresource.WeightedCluster{cluster: {Weight: 1}}}}}}
-	xdsC.InvokeWatchRouteConfigCallback("", xdsresource.RouteConfigUpdate{
-		VirtualHosts: []*xdsresource.VirtualHost{
-			{
-				Domains: []string{targetStr},
-				Routes:  []*xdsresource.Route{{Prefix: newStringP(""), WeightedClusters: map[string]xdsresource.WeightedCluster{cluster: {Weight: 1}}}},
-			},
+// Performs common setup tasks for service watch tests:
+//
+//   - Spins up an xDS management server and configures it with a default
+//     listener and route configuration resource.
+//   - Builds an xDS resolver for the test.
+//   - Verifies that the xDS resolver requests for the expected listener and
+//     route configuration resources.
+//   - Verifies that the xDS resolver pushes the expected service config on the
+//     channel.
+//
+// Returns the following:
+//   - A reference to the xDS management server
+//   - A reference to the test resolver.ClientConn to verify service config pushed
+//     on the channel
+//   - A channel to read requested route configuration resource names from
+func setupForServiceWatchTests(ctx context.Context, t *testing.T, nodeID string) (*e2e.ManagementServer, *testutils.ResolverClientConn, chan []string) {
+	t.Helper()
+
+	listenerResourceNamesCh := make(chan []string, 1)
+	routeConfigResourceNamesCh := make(chan []string, 1)
+
+	// Setup the management server to push the requested listener and route
+	// configuration resource names on to separate channels for the test to
+	// inspect.
+	mgmtServer, err := e2e.StartManagementServer(e2e.ManagementServerOptions{
+		OnStreamRequest: func(_ int64, req *v3discoverypb.DiscoveryRequest) error {
+			switch req.GetTypeUrl() {
+			case version.V3ListenerURL:
+				select {
+				case <-listenerResourceNamesCh:
```

Comment:
Does dropping old updates that haven't been read yet lead to any wonky/funky assertions?